### PR TITLE
Support .offset and .length for dynamic calldata arrays

### LINKED ISSUE #8917: Access to calldata variables of reference type in inline assembly.
Came up in https://github.com/ethereum/solidity/issues/1803
We probably need a similar mechanism as for storage slot/offset for this, only with ``_offset`` and ``_length`` (if it's a dynamic array).
Something like the following should pass:
```
contract C {
  function g(uint256[] calldata x, uint256 index) external returns(uint256 val) {
    assembly {
      switch lt(val, x_length)
      case 0 {
         val := 0xFF
      }
      default {
        val := calldataload(add(x_offset, mul(index, 32)))
      }
    }
  }
  function f() public returns (bool) {
    uint256[] memory x = new uint256[](3);
    x[0] = 0x41; x[1] = 0x42; x[2] = 0x43;
    assert(this.g(x,0) == 0x41);
    assert(this.g(x,1) == 0x42);
    assert(this.g(x,2) == 0x43);
    assert(this.g(x,3) == 0xFF);
    return true;
  }
}
// ----
// f() -> true
```

``function f(uint256[3] calldata x) public { assembly { let y := x_length } }`` or ``struct S { ... } function f(S calldata x) public { assembly { let y := x_length } }`` should be TypeErrors (``_length`` is only present for dynamically sized types).
Regarding name collisions the behaviour should be the same as for ``_offset`` and ``_slot`` for ``storage`` variables.

### LINKED ISSUE #9762: Find Position of Calldata Variable
Currently there is no way to find the starting position of a calldata variable in calldata. This prevents easily parsing calldata with `calldataload`.

Example:
```Solidity
function myfunction(bytes[] calldata info) {
        for(uint i; i < info.length; i++) {
            bytes calldata item = info[i];
            // No way to get the starting position of item in order to
            // parse it with calldataload
        }
    }
```
You might ask why not just use array slices and `abi.decode`.  The answer is that the calldata might not be ABI encoded.  It might be using `abi.encodePacked` and so `abi.decode` cannot be used to decode it.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #9762 
fixes https://github.com/ethereum/solidity/issues/8917

This also fixes a bug in sol->yul about assigning to `.slot`.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: I'd rather go with `suffix != "offset" && suffix != "length"` to avoid the problem of adding new suffixes and forgetting to update these checks. Same applies for the above case with `suffix == length`.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: This one will fire in that case: https://github.com/ethereum/solidity/pull/10208/files#diff-1286fd2f351e6a40b993189fde18c5db99ddfe4a35f4ada894945ba62819a324R794

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: I can add another assertion here if you like.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Oh now I see! Yes, will change.

File: docs/assembly.rst
Comment (hrkrshnn) [inline docs/assembly.rst]:
```suggestion
For dynamic calldata arrays, you can access
```

File: libsolidity/analysis/ReferencesResolver.cpp
Comment (hrkrshnn) [inline libsolidity/analysis/ReferencesResolver.cpp:218]:
```suggestion
	static set<string> suffixes{"length", "offset", "slot"};
```
Maybe have it sorted?

File: libsolidity/ast/ASTAnnotations.h
Comment (hrkrshnn) [inline libsolidity/ast/ASTAnnotations.h:202]: Would `optional<string>` be a better choice?

File: libsolidity/analysis/TypeChecker.cpp
Comment (hrkrshnn) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
			m_errorReporter.typeError(7944_error, _identifier.location, "The suffixes ``.offset``, ``.slot`` and ``.length`` can only be used with variables.");
```
For consistency with https://github.com/ethereum/solidity/blob/571e1e5269472a614148ffd4cd576d9552991671/libsolidity/analysis/ReferencesResolver.cpp#L258

File: libsolidity/ast/ASTJsonConverter.cpp
Comment (hrkrshnn) [inline libsolidity/ast/ASTJsonConverter.cpp]: Can these be removed now?

File: test/libsolidity/ASTJSON/assembly/switch.json
Comment (hrkrshnn) [inline test/libsolidity/ASTJSON/assembly/switch.json]: I think this should be `null` and therefore, not present in the json.

**Review by hrkrshnn — COMMENTED** [review]
Minor comments. Will go through the rest of the code later.

File: libsolidity/analysis/ReferencesResolver.cpp
Comment (hrkrshnn) [reply libsolidity/analysis/ReferencesResolver.cpp:218]: Sorted it not better, I think. According to [this](https://www.cs.usfca.edu/~galles/visualization/RedBlack.html). "offset", "length", "slot" would minimize reordering. But let's ignore this.

File: libsolidity/ast/ASTAnnotations.h
Comment (chriseth) [reply libsolidity/ast/ASTAnnotations.h:202]: Then we would have to deal with a present, but empty string. I mean this way, we cannot really represent `x.`, but I hope that's not a problem...

File: libsolidity/ast/ASTJsonConverter.cpp
Comment (chriseth) [reply libsolidity/ast/ASTJsonConverter.cpp]: I think we should keep them for backwards-compatibility.

File: test/libsolidity/ASTJSON/assembly/switch.json
Comment (chriseth) [reply test/libsolidity/ASTJSON/assembly/switch.json]: good point!

File: test/libsolidity/semanticTests/inlineAssembly/slot_access.sol
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/inlineAssembly/slot_access.sol:29]: Why does this work now?

File: test/libsolidity/syntaxTests/inlineAssembly/two_stack_slots_calldata.sol
Comment (axic) [inline test/libsolidity/syntaxTests/inlineAssembly/two_stack_slots_calldata.sol]: How is this test calldata specific?

File: Changelog.md
Comment (axic) [inline Changelog.md:5]: What is happening on assignment?

File: test/libsolidity/ASTJSON/assembly/slot_offset.json
Comment (axic) [inline test/libsolidity/ASTJSON/assembly/slot_offset.json]: So `suffix: "offset"` but `isOffset: false`?

File: libsolidity/ast/ASTJsonConverter.cpp
Comment (axic) [inline libsolidity/ast/ASTJsonConverter.cpp]:
```suggestion
	tuple["isOffset"] = Json::Value(_info.second.suffix == "offset");
```

File: test/libsolidity/syntaxTests/inlineAssembly/two_stack_slots_calldata.sol
Comment (chriseth) [reply test/libsolidity/syntaxTests/inlineAssembly/two_stack_slots_calldata.sol]: I think I added this test to trigger error nr 9857 - will rename the file.

File: Changelog.md
Comment (chriseth) [reply Changelog.md:5]: As always, it just resets the stack slot contents - just like with `x.slot` for storage pointers.

File: libsolidity/analysis/TypeChecker.cpp
Comment (hrkrshnn) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
					m_errorReporter.typeError(3622_error, _identifier.location, "The suffix .\"" + suffix + "\" is not supported by this variable or type.");
```

File: test/libsolidity/semanticTests/inlineAssembly/calldata_offset_read.sol
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/inlineAssembly/calldata_offset_read.sol:12]: Can you add a test where we have multiple assembly blocks, to see if `CopyTranslate` works properly?

```
    function f(uint, bytes calldata x, uint) public returns (uint r, uint v) {
        assembly {
           x.offset = 8
           x.length = 20
        }
        assembly {
            r := x.offset
            v := x.length
        }
    }
```

**Comment by chriseth** [comment]
Updated the tests.

**Review by hrkrshnn — COMMENTED** [review]
Minor comments. A second review would be nice before merging.

File: libsolidity/analysis/TypeChecker.cpp
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: There is a `.` - I think `."offset"` would sound weird. Note that only `offset`, `slot`, and `length` can appear here.

File: test/libsolidity/semanticTests/inlineAssembly/slot_access.sol
Comment (chriseth) [reply test/libsolidity/semanticTests/inlineAssembly/slot_access.sol:29]: The old implementation did not properly transform the code. It was necessary to do it both in `operator()(yul::Identifier const&` and in `translate(yul::Identifier`.

File: test/libsolidity/semanticTests/inlineAssembly/calldata_offset_read.sol
Comment (chriseth) [reply test/libsolidity/semanticTests/inlineAssembly/calldata_offset_read.sol:12]: Added: https://github.com/ethereum/solidity/pull/10208/files#diff-8d33113ea3ed6f5942382a4d4abf10b353a73ebaf1c224303539522070a4cba9

File: test/libsolidity/syntaxTests/inlineAssembly/storage_reference_on_function.sol
Comment (axic) [inline test/libsolidity/syntaxTests/inlineAssembly/storage_reference_on_function.sol]: When did we start using backticks? The other error message above uses quotes, and the one below uses nothing.

**Comment by chriseth** [comment]
Used moar quotes.

File: test/libsolidity/syntaxTests/inlineAssembly/two_stack_slots_calldata.sol
Comment (axic) [reply test/libsolidity/syntaxTests/inlineAssembly/two_stack_slots_calldata.sol]: Did you rename the file?

**Comment by axic** [comment]
> This also fixes a bug in sol->yul about assigning to .slot.

Should that be mentioned in the changelog? Is there a specific test affected by this?

File: Changelog.md
Comment (axic) [reply Changelog.md:5]: So the use case here would be that functions can modify to which part of the calldata something points to before passing those calldata variables further down.

Perhaps this could be useful for some proxy contracts removing some incoming wrapping?

**Comment by chriseth** [comment]
@axic no, it's a sol->yul bugfix, so no mention in the changelog.

File: Changelog.md
Comment (chriseth) [reply Changelog.md:5]: Another use case is to do something like
```
function sig() returns (bytes calldata s) {
assembly { s.offset := 0 s.length := 4 }
}
```

**Comment by chriseth** [comment]
Renamed the test.

File: libsolidity/codegen/ContractCompiler.cpp
Comment (mijovic) [inline libsolidity/codegen/ContractCompiler.cpp:748]: Do you think it was easier to do it as a inline method `isSlot()` that would do actual check `suffix == "slot"`?

I am fine with leaving as is in this PR. Can be done as a separate refactor if you think it is worth doing it.

Comment (mijovic) [reply libsolidity/codegen/ContractCompiler.cpp:748]: Actually it is not used that much, so probably not even worth it

**Review by mijovic — APPROVED** [review]
I think all comments from @axic and @hrkrshnn are fixed, and this looks good to merge now.

File: libsolidity/codegen/ContractCompiler.cpp
Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp:748]: In the end, these suffixes are exactly the same as the ones in `makeStackItems` in Types.cpp. So maybe at some point, we can just check whether or not they are available and thus support all of them.

**Comment by CryptoKiddies** [comment]
@mijovic @chriseth just want to bring to attention that inline `.offset` behaves differently than offset as described in ABI specification section of docs, which leads to confusion (at least for me). Inline offset points at data, bypassing length prefix, whereas in calldata, the encoded offset points to the length prefix if applicable, as explicitly described in docs. I know this is way belated!

**Pull request merged**
